xapian-core/cluster/cluster.cc

```diff
@@ -644,12 +637,17 @@ StemStopper::add(const string &term)
 {
     LOGCALL_VOID(API, "StemStopper::add", term);
     const string &stemmed_term = 'Z' + stemmer(term);
-    if (stem_action == STEM_NONE)
-	stop_words.insert(term);
-    else if (stem_action == STEM_ALL_Z)
-	stop_words.insert(stemmed_term);
-    else {
-	stop_words.insert(term);
-	stop_words.insert(stemmed_term);
+    switch (stem_action) {
+	case STEM_NONE:
+	    stop_words.insert(term);
+	    break;
+	case STEM_ALL_Z:
+	case STEM_ALL:
```

Comment:
I don't think these should be handled the same way. `STEM_ALL_Z` will generate `Z<stem>` terms, but `STEM_ALL` will generate `<stem>` terms only. So `STEM_ALL` should `stop_words.insert(term)` not `stemmed_term` — I believe.